llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -0,0 +1,286 @@
+static cl::opt<std::string> OutputTxtFilesList{
+    "txt-files-list", cl::desc("Specify output filename for txt files list"),
+    cl::value_desc("filename"), cl::init("-"), cl::cat(ExtractCat)};
+
+static cl::opt<bool> Force{"f", cl::desc("Enable binary output on terminals"),
+                           cl::cat(ExtractCat)};
+
+static cl::opt<bool> OutputAssembly{"S",
+                                    cl::desc("Write output as LLVM assembly"),
+                                    cl::Hidden, cl::cat(ExtractCat)};
+
+static cl::opt<bool> OneKernelPerModule{
+    "one-kernel", cl::desc("Emit a separate module for each kernel"),
+    cl::init(false), cl::cat(ExtractCat)};
+
+static void error(const Twine &Msg) {
+  errs() << "sycl-post-link: " << Msg << '\n';
+  exit(1);
+}
+
+static void error(std::error_code EC, const Twine &Prefix) {
+  if (EC)
+    error(Prefix + ": " + EC.message());
+}
+
+static void writeToFile(std::string Filename, std::string Content) {
+  std::error_code EC;
+  raw_fd_ostream OS{Filename, EC, sys::fs::OpenFlags::OF_None};
+  error(EC, "error opening the file '" + Filename + "'");
+  OS.write(Content.data(), Content.size());
+  OS.close();
+}
+
+static void collectKernelsSet(
+    Module &M, std::map<std::string, std::vector<Function *>> &ResKernelsSet) {
+  for (auto &F : M.functions()) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
+      if (OneKernelPerModule) {
+        ResKernelsSet[F.getName()].push_back(&F);
+      } else if (F.hasFnAttribute("sycl-module-id")) {
+        auto Id = F.getFnAttribute("sycl-module-id");
+        auto Val = Id.getValueAsString();
+        ResKernelsSet[Val].push_back(&F);
+      }
+    }
+  }
+}
+
+// Splits input LLVM IR module M into smaller ones.
+// Input parameter KernelsSet is a map containing groups of kernels with same
+// values in the sycl-module-id attribute. For each group of kernels a separate
+// IR module will be produced.
+// ResModules and ResSymbolsLists are output parameters.
+// Result modules are stored into
+// ResModules vector. For each result module set of kernel names is collected.
+// Sets of kernel names are stored into ResSymbolsLists.
+static void
+splitModule(Module &M,
+            std::map<std::string, std::vector<Function *>> &KernelsSet,
+            std::vector<std::unique_ptr<Module>> &ResModules,
+            std::vector<std::string> &ResSymbolsLists) {
+  for (auto &It : KernelsSet) {
+    // For each group of kernels collect all dependencies.
+    SetVector<GlobalValue *> GVs;
+    std::vector<llvm::Function *> Workqueue;
+    std::string SymbolsList;
+
+    for (auto &F : It.second) {
+      GVs.insert(F);
+      Workqueue.push_back(F);
+      SymbolsList =
+          (Twine(SymbolsList) + Twine(F->getName()) + Twine("\n")).str();
+    }
+
+    while (!Workqueue.empty()) {
+      Function *F = &*Workqueue.back();
+      Workqueue.pop_back();
+      for (inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
```

Comment:
I am unsure this change makes this code beautiful... :-)
What about https://github.com/triSYCL/sycl/pull/71/files#diff-e20cbcc5e97c60007cd492e6146e2210R407 instead?

Reply:
Déjà vu...

Reply:
Using inst iterator is recommennded by LLVM manual:
http://releases.llvm.org/2.9/docs/ProgrammersManual.html#iterate_institer

Reply:
... but what @keryell suggests looks indeed more elegant inded:
```for (auto &I : instructions(F))```

Reply:
Definitely more elegant, thank you!